README
LinkLocal-IPv4-Interface version 0.15
=====================================

This package represents a lightweight, pure Perl implementation of the specification
as outlined in RFC-3927, Dynamic Configuration of IPv4 Link-Local Addresses.
This standard details a mechanism which provides a means of enabling automatic IPv4
address allocations to network addressable entities and devices which by happenstance
or design find themselves on an unmanaged, ad hoc IP network with no other means of
acquiring a fully routable, unicast IPv4 address. As the name implies, link-local IPv4
addresses are non-routable with all participating devices connected to the same physical
link; this fact all but dictates that such networks are typically small. Home networks or
other such small environments are well suited to the deployment of this standard. While 
there are, in theory, upwards of 65,024 addresses available in the range for allocations,
the reality is that as more addresses from the reserved range are allocated to hosts on
the local link the number of collisions necessarily increases as well which degrades 
overall performance. Products such as QIP meet this need in the enterprise.

This standard is widely implemented in most major operating systems, most of which "more 
or less" adhere to this specification. While I have remained true to the specification
and to the algorithms detailed therein I have in some ways made "enhancements"; one
example being my pseudorandom address selection will generate link-local addresses in
blocks of ten. This is similar to the implementation found in OS X and some of the 
various offerings by Microsoft. This decision was simply a matter of convenience and 
nothing more.

This module, LinkLocal::IPv4::Interface provides the main entry point into this
implementation. Because the specification recommends that link-local address are generated
in a pseudorandom fashion using something unique to the particular host on which they
are being generated such as a MAC address, I have by design coupled the selection of 
addresses closely with the interface object itself; given that I am indeed using the 
interface hardware MAC as a seed to srand() this just made good sense to me. The primary
gateway into this process is the call to the if_config() method. This method will step
through the various steps in the address selection process and will configure the interface
with a fully compliant IPv4 link-local address upon completion.

There are some things I would like to note here. First, this implementation makes absolutely
no assumptions about the current state of the interface at any time. Client code should
know this. The specification is quite clear that IPv4 link-local addresses are to be used
only in the absence of a fully routable IPv4 address. The dynamic configuration of an
IPv4 link-local address is to never interfere with the operation of any of the other
processes whereby a host can acquire a standard, routable address such as a DHCP state
machine or static assignments under the direction of a network administrator. This 
implementation is only to be used at such a time that a determination has been made that
there are no operable and routable addresses configured on the interface. This library
does not at any time make any assumptions about the operability of any addresses assigned
to any interfaces on the host, nor does it in any way attempt to determine the management
status of any network in which it participates. This is an important point to consider here 
as small networks have become ubiquitous and "network hopping" with portable devices is 
certainly a scenario that is much more common today than it was ten or even five years ago.
Secondly, the configuration of an IPv4 link-local address is not something which is a
static thing; the implementation does not just probe the local physical link for the
address, configure the interface and forget about it. At any time an address conflict may
occur at which time the selection process may begin again. I have addressed this fact in my
design in a broad way. Client code will be expected to register a callback on a hook I 
provide. Furthermore, I will also expose an optional pure Perl eventing mechanism which can be
set to monitor ARP on the local link for any requests or replies which contain the currently
configured link-local address. The other option here is that client code may wish to bring
its own eventing mechanism to the party. As such I have met both needs by way of Marc
Lehmann's AnyEvent module which has been described as the DBI of event loops. For those
not familiar with Marc's AnyEvent and Coro material, you might want to run over and check
out his stuff. In the world of asynchronous Perl eventing and coroutines, Marc is Master Yoda.

NOTE: This is still very much a work in progress. Much of this is pulled from some old 
stuff which I had around, including bits and bobs of some UPnP stuff I had in C and which
I reimplemented in Perl. As such, some design is happening in situ. In addition to this, I
do, unfortunately, have to go to work and feed myself so I am unable to hack on this and
other things 24/7. I decided that I would push to CPAN early and chase my crufty push rather
than keep it in a git repo on my Mac and constantly pick and preen it and never get anything
actually out. So all that said, I will release as quickly and as often as I can and hopefully
I will have something here approaching actually useful before long at all, so please bear
with me and thanks for your patience.

Also, this is fully intended to work along side such other protocols as Multicast Service
Discovery and other lightweight service announcement protocols. The ability to get a
valid address usable in communication with other hosts and devices on an unmanaged network
is not terribly useful if you cannot find anything out about the particular network in 
question. As such I am toying with the idea of integrating, albeit very loosely, one or more
of the service discovery mechanisms out there as this is what makes link-local addressing
actually useful, scenarios such as plugging a laptop and a printer into a network and they
just work. While yes, there are full zero configuration suites out there in C and in other
languages, there is no lightweight equivalent in Perl.... yet.


INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  Moose
  Moose::Util::TypeConstraints
  IO::Interface::Simple
  Net::Frame::Layer::Arp
  Regexp::Common

COPYRIGHT AND LICENCE

Copyright (C) 2010 Ray Mroz

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.




Changes
commit a07df83cc464ec826dd867506e3efbb185b0037e
Author: Ray Mroz <raymondmroz@gmail.com>
Date:   Tue Nov 30 22:05:55 2010 -0400

    Removed some debug prints, two includes and updated POD.

 lib/LinkLocal/IPv4/Interface/Cache.pm |   13 ++++---------
 lib/LinkLocal/IPv4/Interface/Types.pm |    2 +-
 2 files changed, 5 insertions(+), 10 deletions(-)

commit 94f9fc205ede318739455e1115d90b258252e54f
Author: Tony Li Xu <tonylixu@gmail.com>
Date:   Tue Nov 30 21:41:33 2010 -0400

    Perltidy ran on test.pl and Cache.pm

 bin/test.pl                           |    4 +-
 lib/LinkLocal/IPv4/Interface/Cache.pm |  124 +++++++++++++++++----------------
 2 files changed, 66 insertions(+), 62 deletions(-)

commit f3991f1c4ad81dd69628dfde64d5d5d0229b132a
Author: Tony Li Xu <tonylixu@gmail.com>
Date:   Tue Nov 30 21:26:13 2010 -0400

    Update Cache.pm and test.pl, finished functions in Cache.pn

 bin/test.pl                           |   19 +--
 lib/LinkLocal/IPv4/Interface/Cache.pm |  200 +++++++++++++++++++--------------
 2 files changed, 121 insertions(+), 98 deletions(-)

commit 4f91731f4b296024123e699346e8e5b3dfa153c8
Author: Ray Mroz <raymondmroz@gmail.com>
Date:   Tue Nov 30 20:47:26 2010 -0400

    Cleaned up POD in several files.

 bin/test.pl                               |   18 +++-
 lib/LinkLocal/IPv4/Interface.pm           |   12 ++--
 lib/LinkLocal/IPv4/Interface/Cache.pm     |  135 ++++++++++++++---------------
 lib/LinkLocal/IPv4/Interface/Constants.pm |    6 +-
 lib/LinkLocal/IPv4/Interface/Types.pm     |    6 +-
 5 files changed, 92 insertions(+), 85 deletions(-)

commit 86dd443e0ec4dbaececdb9c9e02f0af0f6e59147
Author: Ray Mroz <raymondmroz@gmail.com>
Date:   Mon Nov 29 17:55:41 2010 -0400

    Added Cache.pm, partially completed.
    
    Fixed some POD in other modules.

 README                                    |   86 +++++++++-
 bin/test.pl                               |   14 +-
 lib/LinkLocal/IPv4/Interface.pm           |   22 ++--
 lib/LinkLocal/IPv4/Interface/ARP.pm       |    5 +
 lib/LinkLocal/IPv4/Interface/Cache.pm     |  272 +++++++++++++++++++++++++++++
 lib/LinkLocal/IPv4/Interface/Constants.pm |   62 +++++++-
 6 files changed, 438 insertions(+), 23 deletions(-)

commit d80c95bfed728f9cca2fb9b7351bf62029e941b9
Author: Ray Mroz <raymondmroz@gmail.com>
Date:   Wed Nov 24 19:19:06 2010 -0400

    Cleaned up all POD

 bin/test.pl                               |    4 +
 lib/LinkLocal/IPv4/Interface.pm           |  202 +++++++++++++++++++++++++++--
 lib/LinkLocal/IPv4/Interface/ARP.pm       |   88 +++++++------
 lib/LinkLocal/IPv4/Interface/Constants.pm |    2 +-
 lib/LinkLocal/IPv4/Interface/Types.pm     |   89 ++++++++-----
 5 files changed, 300 insertions(+), 85 deletions(-)

commit 2439fa2191590724f4dc5aa269201159b28da101
Author: Ray Mroz <raymondmroz@gmail.com>
Date:   Tue Nov 23 23:57:21 2010 -0400

    Some small file tweaks, moved some additional modules to the Types.pm file

 lib/LinkLocal/IPv4/Interface.pm       |    2 --
 lib/LinkLocal/IPv4/Interface/ARP.pm   |    1 -
 lib/LinkLocal/IPv4/Interface/Types.pm |    3 +++
 3 files changed, 3 insertions(+), 3 deletions(-)

commit 04d50108724d2d1db10888bbb3d0e86caf833cbd
Author: Ray Mroz <raymondmroz@gmail.com>
Date:   Tue Nov 23 23:51:05 2010 -0400

    Added ARP.pm, Types.pm.  Completed POD for both types.
    
    ARP.pm - this Moose class represents an object wrapper around ARP
    packet messages which are required by the specification, specifically
    ARP PROBE and ARP ANNOUNCE.  Please see RFC-3927 for details.
    Types.pm - this is just a collection of shared custom types which will
    be included in files package wide.

 bin/test.pl                               |   12 ++-
 lib/LinkLocal/IPv4/Interface.pm           |   29 ++----
 lib/LinkLocal/IPv4/Interface/ARP.pm       |  182 +++++++++++++++++++++++++++++
 lib/LinkLocal/IPv4/Interface/Constants.pm |   17 +++-
 lib/LinkLocal/IPv4/Interface/Types.pm     |  149 +++++++++++++++++++++++
 5 files changed, 365 insertions(+), 24 deletions(-)

commit b3a41622615816f1332e6134939bf1bf0c6caabb
Author: Ray Mroz <raymondmroz@gmail.com>
Date:   Mon Nov 22 19:28:47 2010 -0400

    Removed .TextMate project file from repository

 LinkLocal.tmproj |   16 ----------------
 1 files changed, 0 insertions(+), 16 deletions(-)

commit 665b49c3f6af98f7f5ddb0d4e54846678cfa6e9e
Author: Ray Mroz <raymondmroz@gmail.com>
Date:   Mon Nov 22 19:27:00 2010 -0400

    Removed .DS_Store OS X specific files from repository

 .DS_Store               |  Bin 6148 -> 0 bytes
 lib/.DS_Store           |  Bin 6148 -> 0 bytes
 lib/LinkLocal/.DS_Store |  Bin 6148 -> 0 bytes
 3 files changed, 0 insertions(+), 0 deletions(-)

commit 02f8a6fc3348218b2d2d1ad81672288444debc2f
Author: Ray Mroz <raymondmroz@gmail.com>
Date:   Mon Nov 22 00:05:53 2010 -0400

    Pushing changes to reflect CPAN upload

 README                          |   36 ++++++++++++++++++++----------------
 lib/LinkLocal/IPv4/Interface.pm |   32 ++++++++++++++++++++------------
 rfc3927.txt                     |    6 ------
 3 files changed, 40 insertions(+), 34 deletions(-)

commit da994b999b2cd71c33acad7d11a01ea8619427fa
Author: Ray Mroz <raymondmroz@gmail.com>
Date:   Sun Nov 21 23:17:38 2010 -0400

    Added copy of rfc3927 text

 rfc3927.txt | 1851 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 1 files changed, 1851 insertions(+), 0 deletions(-)

commit 1df0b2fbd5543c70d27c467bf15ddceb4b59b4fd
Author: Ray Mroz <raymondmroz@gmail.com>
Date:   Sun Nov 21 22:55:49 2010 -0400

    Additional cleaning, preperation for CPAN

 bin/test.pl                     |    8 +++++++-
 lib/LinkLocal/IPv4/Interface.pm |   25 +++++++++++++------------
 2 files changed, 20 insertions(+), 13 deletions(-)

commit a92323a23dd00f217781138fc6d8aadf41255d07
Author: Ray Mroz <raymondmroz@gmail.com>
Date:   Sun Nov 21 22:03:57 2010 -0400

    Cleaned up some testing cruft

 bin/test.pl                     |   33 +++++++++++++

lib/LinkLocal/IPv4/Interface/ARP.pm
package LinkLocal::IPv4::Interface::ARP;

require 5.010000;

# Copyright (C) 2010 Raymond Mroz
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

use Moose;
use Moose::Util::TypeConstraints;
use LinkLocal::IPv4::Interface::Types;

# ==========
# = packet =
# ==========
has 'packet' => (
    is      => 'rw',
    isa     => 'ArpPacket',
    handles => {
        pack_arp        => 'pack',
        unpack_arp      => 'unpack',
        get_key         => 'getKey',
        get_key_reverse => 'getKeyReverse',
        is_match        => 'match',
        print_arp       => 'print',
        dump_arp        => 'dump',
    },
    coerce => 1,
);

sub send {
    my $this = shift;
    # TODO Handle send, packet object will be passed into a "sender machine"
}

no Moose;

__PACKAGE__->meta->make_immutable;
1;

=head1 NAME

LinkLocal::IPv4::Interface::ARP - An ARP packet wrapper, this class allows for the easy construction
of either Probe or Announce ARP Requests, both of which are central to dynamic link-local address
configuration.

=head1 SYNOPSIS

  use LinkLocal::IPv4::Interface::ARP;

  $arp = LinkLocal::IPv4::Interface::ARP->new();

  # Initialize a hash reference for each type of ARP message required.
  my $set_probe    = { srcIp => '000.000.000.000', dstIp => '169.254.100.100' };
  my $set_announce = { srcIp => '169.254.100.100', dstIp => '169.254.100.100' };

  # Setup an ARP Probe packet, testing if link-local address 169.254.100.100 is in use
  $arp->packet( $set_probe );

  # Setup an ARP Announce packet, claiming link-local address 169.254.100.100 on network
  $arp->packet( $set_announce );
  
  # Also valid, if no srcIp hash key, assumes probe and zeroes sender address
  my $set_probe    = { dstIp => '169.254.100.100' };

=head1 DESCRIPTION

This small ARP packet wrapper class plays a critical role in the dynamic configuration of IPv4
link-local addresses. When a new address is pseudo-randomly determined, ARP Probe messages are
sent out over the interface which is being configured with a link-local address. The ARP packets
are of type ARP Request, have the source address zeroed out and the target address (dstIp) set 
to the address which has been selected and is being probed. Three of these ARP Probes are sent 
at randomly spaced intervals and if no conflict is detected during the process, the implementation 
will claim the address for the interface within the unmanaged network.

After an address has been claimed, the implementation requires that three ARP Announce packets 
be broadcast on the network. These packets have both sender and target addresses set to the 
address which has been claimed by the implementation. This is done to clean up any stale ARP
cache's on the local-link.

=head2 ATTRIBUTES

=over 4

=item C<packet>

This attribute holds the custom Moose type ArpPacket which in essence provides and allows for
the Moose class wrapper around the C<Net::Frame::Layer::ARP> object type. The packet attribute
exposes a number of mappings via the Moose attribute handler construct which allow for direct
mapping between them and the object reference type without forcing the additional layer of
indirection upon client code.

=item C<raw>

This attribute is inherited from the original C<Net::Frame::Layer> class and it holds the raw
layer as has just been pulled from the network or has been packed for network transmission (see 
the dump_arp method below).

=back

=head2 METHODS

=over 4

=item C<packet()>

The packet() method provides the most basic means of setting the target and sender addresses
to the desired value, depending upon the type of ARP packet being constructed. packet() takes
both a blessed or non-blessed hash reference, so the return type from object creation can be
passed to this method as long as this blessed reference isa C<Net::Frame::Layer::ARP> object.

=item C<pack_arp()>

Packs all attributes into a raw format suitable for injection into the network. Returns the 
raw packed string on success, undef otherwise. Result is stored into raw attribute. See 
L<Net::Frame::Layer> for further details.

=item C<unpack_arp()>

Unpacks raw data from network and stores attributes into the object. Returns $this on success, 
undef otherwise.

=item C<get_key()> 

=item C<get_key_reverse()>

These two methods are basically used to increase speed when using the recv method from 
C<Net::Frame::Simple> via the packet object attribute. Usually, you use them when you need 
to write an is_match() method.

=item C<is_match()>

This method is mostly used internally. You pass a C<LinkLocal::IPv4::Interface::ARP> packet 
attribute as a parameter, and it returns true if this is a response corresponds with the 
request, otherwise false.

=item C<print_arp()>

Prints a human readable string representation of the packet and its underlying state.

=item C<dump_arp()>

Prints a hexadecimal formatted respresentation of the packet, exactly how this would appear on
the network.

=back

B<Please Note>: While this class does not directly override or map all of the underlying attributes
and methods which are present in the base C<Net::Frame::Layer> type from which the Moose
class is derived, one could, with an extra level of indirection, access the underlying
attributes and methods from that base type. Clients of this module are encouraged to either conform
to the current interface or to extend it via Moose's inheritance mechanism if they wish to access
additional underlying functionality in this way. Please see the relevant POD for further details.

=head1 SEE ALSO

Refer to RFC-3927, "Dynamic Configuration of IPv4 Link-Local Adresses", the complete
text of which can be found in the top level of the package archive.

L<perl>, L<Net::Frame::Layer>, L<Net::Frame::Layer::ARP>, L<Moose>

This project is also hosted on github at:
	git@github.com:raymroz/LinkLocal--IPv4.git

=head1 AUTHOR

Ray Mroz, E<lt>mroz@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2010 Ray Mroz

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.


=cut


lib/LinkLocal/IPv4/Interface.pm
package LinkLocal::IPv4::Interface;

require 5.010000;

# require Exporter;
# our @ISA         = qw(Exporter);
# our %EXPORT_TAGS = ();
# our @EXPORT_OK   = ();
# our @EXPORT      = qw();

# Copyright (C) 2010 Raymond Mroz
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
# Requires:
#	Moose
#	Moose::Util::TypeConstraints
#	IO::Interface::Simple
#	Regexp::Common

use Moose;
use Moose::Util::TypeConstraints;
use LinkLocal::IPv4::Interface::Types;

# =============
# = interface =
# =============
has 'interface' => (
    is      => 'bare',
    isa     => 'LinkLocalInterface',
    handles => {
        get_if_device => 'name',
        get_if_index  => 'index',
        get_if_list   => 'interfaces',
        get_if_addr   => 'address',
        set_if_addr   => 'address',
        get_if_mac    => 'hwaddr',
    },
    coerce => 1,
);

# ================
# = address_list =
# ================
has 'address_list' => (
    is       => 'ro',
    isa      => 'ArrayRef[IpAddress]',
    reader   => '_get_address_list',
    builder  => '_build_address_list',
    lazy     => 1,
    init_arg => undef,
);

# =============
# = BUILDARGS =
# =============
around BUILDARGS => sub {
    my $orig = shift;
    my $this = shift;

    if ( @_ == 1 && !ref $_[0] ) {
        return $this->$orig( interface => $_[0] );
    }
    else {
        return $this->$orig(@_);
    }
};

# =======================
# = _build_address_list =
# =======================
sub _build_address_list {
    my $this = shift;

    # Generate srand() seed from mac address
    my $mac       = $this->get_if_mac;
    my @split_mac = split( /':'/, $mac );
    my $seed      = 0;
    foreach my $element (@split_mac) {
        $seed += hex($element);
    }

    # Seed for pseudo-random address generation
    srand($seed);

    # Create a list of 10 pseudo-random link-local addresses
    my @llv4_ip_list = ();
    for ( my $iter = 0 ; $iter < 10 ; $iter++ ) {
        $llv4_ip_list[$iter] =
          join( '.', '169.254', ( 1 + int( rand(254) ) ), int( rand(256) ) );
    }

    return \@llv4_ip_list;
}

# ====================
# = get_next_address =
# ====================
sub _get_next_address {
    my $this = shift;

    # TODO Fix this, it will break after 10 pulls
    return shift( @{ $this->_get_address_list() } );
}

# =============
# = if_config =
# =============
sub if_config {
	my $this = shift;
	
	# TODO Call back end code here 
}

no Moose;

__PACKAGE__->meta->make_immutable;
1;

=head1 NAME

LinkLocal::IPv4::Interface - IPv4 link-local interface wrapper

=head1 SYNOPSIS

  use LinkLocal::IPv4::Interface;
  
  my $if_obj = LinkLocal::IPv4::Interface->new('eth0');
  
  # TODO: finalize method signature and process
  $if_obj->if_config( $arg_a, $arg_b ); 

=head1 DESCRIPTION

This package represents a lightweight, pure Perl implementation of the specification
as outlined in RFC-3927, I<Dynamic Configuration of IPv4 Link-Local Addresses>.
This standard details a mechanism which provides a means of enabling automatic IPv4
address allocations to network addressable entities and devices which by happenstance
or design find themselves on an unmanaged, ad hoc IP network with no other means of
acquiring a fully routable, unicast IPv4 address. As the name implies, link-local IPv4
addresses are non-routable in the IANA reserved C<169.254/16> prefix, with all participating
devices connected to the same physical link; this fact all but dictates that such networks 
are typically small. Home networks or other such small environments are well suited to the 
deployment of this standard. While there are, in theory, upwards of 65,024 addresses available
in the range for allocations, the reality is that as more addresses from the reserved range 
are allocated to hosts on the local link the number of collisions necessarily increases as 
well which degrades overall performance. Products such as QIP meet this need in the enterprise.

This standard is widely implemented in most major operating systems, most of which "more 
or less" adhere to this specification. While I have remained true to the specification
and to the algorithms detailed therein I have in some ways made "enhancements"; one
example being my pseudorandom address selection will generate link-local addresses in
blocks of ten. This is similar to the implementation found in OS X and some of the 
various offerings by Microsoft. This decision was simply a matter of convenience and 
nothing more.

This module, C<LinkLocal::IPv4::Interface> provides the main entry point into this
implementation. Because the specification recommends that link-local address are generated
in a pseudorandom fashion using something unique to the particular host on which they
are being generated such as a MAC address, I have by design coupled the selection of 
addresses closely with the interface object itself; given that I am indeed using the 
interface hardware MAC as a seed to C<srand()> this just made good sense to me. The primary
gateway into this process is the call to the C<if_config()> method. This method will step
through the various steps in the address selection process and will configure the interface
with a fully compliant IPv4 link-local address upon completion.

There are some things I would like to note here. First, this implementation makes absolutely
no assumptions about the current state of the interface at any time. Client code should
know this. The specification is quite clear that IPv4 link-local addresses are to be used
only in the absence of a fully routable IPv4 address. The dynamic configuration of an
IPv4 link-local address is to never interfere with the operation of any of the other
processes whereby a host can acquire a standard, routable address such as a DHCP state
machine or static assignments under the direction of a network administrator. This 
implementation is B<only> to be used at such a time that a determination has been made that
there are no operable and routable addresses configured on the interface. This library
does not at any time make any assumptions about the operability of any addresses assigned
to any interfaces on the host, nor does it in any way attempt to determine the management
status of any network in which it participates. This is an important point to consider here 
as small networks have become ubiquitous and "network hopping" with portable devices is 
certainly a scenario that is much more common today than it was ten or even five years ago.
Secondly, the configuration of an IPv4 link-local address is not something which is a
static thing; the implementation does not just probe the local physical link for the
address, configure the interface and forget about it. At any time an address conflict may
occur at which time the selection process may begin again. I have addressed this fact in my
design in a broad way. Client code will be expected to register a callback on a hook I 
provide. Furthermore, I will also expose an optional pure Perl eventing mechanism which can be
set to monitor ARP on the local link for any requests or replies which contain the currently
configured link-local address. The other option here is that client code may wish to bring
its own eventing mechanism to the party. As such I have met both needs by way of Marc
Lehmann's AnyEvent module which has been described as the DBI of event loops. For those
not familiar with Marc's AnyEvent and Coro material, you might want to run over and check
out his stuff. In the world of asynchronous Perl eventing and coroutines, Marc is Master Yoda.

B<NOTE:> This is still very much a work in progress. Much of this is pulled from some old 
stuff which I had around, including bits and bobs of some UPnP stuff I had in C and which
I reimplemented in Perl. As such, some design is happening in situ. In addition to this, I
do, unfortunately, have to go to work and feed myself so I am unable to hack on this and
other things 24/7. I decided that I would push to CPAN early and chase my crufty push rather
than keep it in a git repo on my Mac and constantly pick and preen it and never get anything
actually out. So all that said, I will release as quickly and as often as I can and hopefully
I will have something here approaching actually useful before long at all, so please bear
with me and thanks for your patience.

Also, this is fully intended to work along side such other protocols as Multicast Service
Discovery and other lightweight service announcement protocols. The ability to get a
valid address usable in communication with other hosts and devices on an unmanaged network
is not terribly useful if you cannot find anything out about the particular network in 
question. As such I am toying with the idea of integrating, albeit very loosely, one or more
of the service discovery mechanisms out there as this is what makes link-local addressing
actually useful, scenarios such as plugging a laptop and a printer into a network and they
just work. While yes, there are full zero configuration suites out there in C and in other
languages, there is no lightweight equivalent in Perl.... yet.

=head2 DEPENDENCIES

L<Moose>

L<Moose::Util::TypeConstraints>

L<MooseX::Params::Validate>

L<IO::Interface::Simple>

L<Net::Frame::Layer::ARP>

L<Net::Frame::Simple>

L<Regexp::Common>

=head2 ATTRIBUTES

=over 4

=item C<interface>

The interface attribute contains a reference to the C<IO::Interface::Simple> object
type which is instantiated via a C<Moose> type coercion from a string representation of
the interface device name. The interface attribute also provides for a number of 
mappings to the underlying objects methods and attributes by way of the very powerful
C<Moose> delegation feature.

=item C<address_list>

The address_list attribute is set to contain a list of ten MAC seeded, pseudorandomly
generated IPv4 link-local addresses. As interfaces are checked for collison on the 
network by way of ARP probes to determine if they are already in use, they are popped 
off of the list.

=back

=head2 CONSTRUCTORS

=over 4

=item C<new>

This lone constructor takes the name of a network device on the system, as a string
and returns a Moose-wrapped C<IO::Interface::Simple> object.

=back

=head2 METHODS

=over 4

=item C<get_if_device>

This call returns a string representation of the interface.

=item C<get_if_index>

Returns the interface index (only valid on BSD-style systems).

=item C<get_if_list>

Returns a list comprised of all detected interfaces on the system.  Note, this
will not return a member for the loopback interface.

=item C<get_if_addr>

Returns the string representation of the IPv4 address to which the interface is
currently configured. The address is in dotted-decimal notation.

=item C<set_if_addr>

Attempts to set the interface to the IPv4 address string which it takes as its lone
parameter. The address must be in dotted-decimal format.

=item C<get_if_mac>

Returns the string representation of the hardware address of the interface. The
returned hardware address is in standard colon seperated, hexadecimal digit format.

=item C<get_next_ip>

Returns and pops the next available, pseudo-randomly generated IPv4 link-local address
from the attribute list, shrinking that list by one.

=item C<if_config>

General gateway method to this framework. A call to if_config begins the process of
dynamically configuring the specified interface with an IPv4 link-local address. This 
call initiates the process of checking the link-local address cache for the last successfully
configured address otherwise a list of ten, new pseudo-random generated addresses is
created. The implementation fully adheres to the specification as defined in RFC-3927
as it ARP Probes the first address in the list or from the cache to determine if it is
already in use elsewhere on the local link. If it gets no response to its probe packets,
it configures the interface with the address and sends ARP Announce packets announcing
that it has claimed the address as well as to clear stale ARP cache entries which might
exist out on the link. If at any time an ARP packet, request or reply, is detected which
contains the address it has selected, it will either concede the address or make a single
attempt at address defense, depending upon the context of the packet and its current 
state.

=back

=head2 EXPORT

	PROBE_WAIT          =>  1 second
	PROBE_NUM           =>  3
	PROBE_MIN           =>  1 second
	PROBE_MAX           =>  2 seconds
	ANNOUNCE_WAIT       =>  2 seconds
	ANNOUNCE_NUM        =>  2
	ANNOUNCE_INTERVAL   =>  2 seconds
	MAX_CONFLICTS       => 10
	RATE_LIMIT_INTERVAL => 60 seconds
	DEFEND_INTERVAL     => 10 seconds


=head1 SEE ALSO

Refer to RFC-3927, I<Dynamic Configuration of IPv4 link-local Adresses>, the complete
text of which can be found in the top level of the package archive.

L<perl>, L<IO::Interface::Simple>, L<Moose>

This project is also hosted on github at:
	git@github.com:raymroz/LinkLocal--IPv4.git

=head1 AUTHOR

Ray Mroz, E<lt>mroz@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2010 Ray Mroz

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.


=cut
